File: span-builder/src/main/java/com/maciega/bartosz/span_builder/spans/SpanBuilder.java
package com.maciega.bartosz.span_builder.spans;

import android.annotation.TargetApi;
import android.content.Context;
import android.content.res.ColorStateList;
import android.graphics.Bitmap;
import android.graphics.MaskFilter;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Build;
import android.os.LocaleList;
import android.os.Parcel;
import android.text.Layout;
import android.text.SpannableStringBuilder;
import android.text.Spanned;
import android.text.method.LinkMovementMethod;
import android.view.View;
import android.widget.TextView;

import com.maciega.bartosz.span_builder.spans.builders.AbsoluteSizeSpanBuilder;
import com.maciega.bartosz.span_builder.spans.builders.BackgroundColorSpanBuilder;
import com.maciega.bartosz.span_builder.spans.builders.ClickableSpanBuilder;
import com.maciega.bartosz.span_builder.spans.builders.DrawableMarginSpanBuilder;
import com.maciega.bartosz.span_builder.spans.builders.ForegroundColorSpanBuilder;
import com.maciega.bartosz.span_builder.spans.builders.IconMarginSpanBuilder;
import com.maciega.bartosz.span_builder.spans.builders.IconSpanBuilder;
import com.maciega.bartosz.span_builder.spans.builders.ImageSpanBuilder;
import com.maciega.bartosz.span_builder.spans.builders.MaskFilterSpanBuilder;
import com.maciega.bartosz.span_builder.spans.builders.StyleSpanBuilder;
import com.maciega.bartosz.span_builder.spans.builders.SuggestionSpanBuilder;
import com.maciega.bartosz.span_builder.spans.builders.TextAppearanceSpanBuilder;
import com.maciega.bartosz.span_builder.spans.builders.AlignmentSpanStandardBuilder;
import com.maciega.bartosz.span_builder.spans.builders.BulletSpanBuilder;
import com.maciega.bartosz.span_builder.spans.builders.LeadingMarginStandardSpanBuilder;
import com.maciega.bartosz.span_builder.spans.builders.LocaleSpanBuilder;
import com.maciega.bartosz.span_builder.spans.builders.QuoteSpanBuilder;
import com.maciega.bartosz.span_builder.spans.builders.RelativeSizeSpanBuilder;
import com.maciega.bartosz.span_builder.spans.builders.ScaleXSpanBuilder;
import com.maciega.bartosz.span_builder.spans.builders.StrikeThroughSpanBuilder;
import com.maciega.bartosz.span_builder.spans.builders.TabStopStandardSpanBuilder;
import com.maciega.bartosz.span_builder.spans.builders.TypeFaceSpanBuilder;
import com.maciega.bartosz.span_builder.spans.builders.UnderlineSpanBuilder;
import com.maciega.bartosz.span_builder.spans.builders.UrlSpanBuilder;

import java.util.Locale;

/**
 * Created by bartoszmaciega on 28/02/17.
 */

/**
 * Things to do:
 * - route spans to helper classes( separated by superclasses)
 * - rethink if should remove constructors with parcel
 */

public class SpanBuilder {
    private SpannableStringBuilder spannableBuilder;
    private String text;
    private TextView textView;
    private int startIndex;
    private int endIndex;
    private int flags;


    public static SpanBuilder newInstance(String text) {
        return new SpanBuilder(text);
    }


    SpannableStringBuilder getSpannableBuilder() {
        return spannableBuilder;
    }

    private SpanBuilder(String text) {
        this.text = text;
        spannableBuilder = new SpannableStringBuilder(text);
        startIndex = 0;
        endIndex = text.length();
        flags = Spanned.SPAN_INCLUSIVE_EXCLUSIVE;
    }

    public SpanBuilder withBackgroundColor(int color) {
        BackgroundColorSpanBuilder builder = new BackgroundColorSpanBuilder(color);
        return makeSpan(builder);
    }

    public SpanBuilder withBackgroundColor(Parcel src) {
        BackgroundColorSpanBuilder builder = new BackgroundColorSpanBuilder(src);
        return makeSpan(builder);
    }

    public SpanBuilder withForegroundColor(int color) {
        ForegroundColorSpanBuilder builder = new ForegroundColorSpanBuilder(color);
        return makeSpan(builder);
    }

    public SpanBuilder withForegroundColor(Parcel src) {
        ForegroundColorSpanBuilder builder = new ForegroundColorSpanBuilder(src);
        return makeSpan(builder);
    }

    public SpanBuilder withClickable(View.OnClickListener listener) {
        ClickableSpanBuilder builder = new ClickableSpanBuilder(listener);
        if (textView != null) {
            textView.setMovementMethod(LinkMovementMethod.getInstance());
        }
        return makeSpan(builder);
    }

    public SpanBuilder withUrl(String url) {
        UrlSpanBuilder builder = new UrlSpanBuilder(url);
        if (textView != null) {
            textView.setMovementMethod(LinkMovementMethod.getInstance());
        }
        return makeSpan(builder);
    }

    public SpanBuilder withMask(MaskFilter mask) {
        MaskFilterSpanBuilder builder = new MaskFilterSpanBuilder(mask);
        return makeSpan(builder);
    }

    public SpanBuilder withAbsoluteSize(int size) {
        AbsoluteSizeSpanBuilder builder = new AbsoluteSizeSpanBuilder(size, false);
        return makeSpan(builder);
    }

    public SpanBuilder withAbsoluteSize(int size, boolean dip) {
        AbsoluteSizeSpanBuilder builder = new AbsoluteSizeSpanBuilder(size, dip);
        return makeSpan(builder);
    }

    public SpanBuilder withAbsoluteSize(Parcel src) {
        AbsoluteSizeSpanBuilder builder = new AbsoluteSizeSpanBuilder(src);
        return makeSpan(builder);
    }

    @TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR1)
    public SpanBuilder withLocale(Locale locale) {
        LocaleSpanBuilder builder = new LocaleSpanBuilder(locale);
        return makeSpan(builder);
    }

    @TargetApi(Build.VERSION_CODES.N)
    public SpanBuilder withLocale(LocaleList localeList) {
        LocaleSpanBuilder builder = new LocaleSpanBuilder(localeList);
        return makeSpan(builder);
    }

    public SpanBuilder withLocale(Parcel src) {
        LocaleSpanBuilder builder = new LocaleSpanBuilder(src);
        return makeSpan(builder);
    }

    public SpanBuilder withRelative(float proportion) {
        RelativeSizeSpanBuilder builder = new RelativeSizeSpanBuilder(proportion);
        return makeSpan(builder);
    }

    public SpanBuilder withRelative(Parcel src) {
        RelativeSizeSpanBuilder builder = new RelativeSizeSpanBuilder(src);
        return makeSpan(builder);
    }

    public SpanBuilder withImage(Context context, Bitmap bitmap) {
        ImageSpanBuilder builder = new ImageSpanBuilder(context, bitmap);
        return makeSpan(builder);
    }

    public SpanBuilder withImage(Context context, Bitmap bitmap, int verticalAlignment) {
        ImageSpanBuilder builder = new ImageSpanBuilder(context, bitmap, verticalAlignment);
        return makeSpan(builder);
    }

    public SpanBuilder withImage(Drawable drawable) {
        ImageSpanBuilder builder = new ImageSpanBuilder(drawable);
        return makeSpan(builder);
    }

    public SpanBuilder withImage(Drawable drawable, String source) {
        ImageSpanBuilder builder = new ImageSpanBuilder(drawable, source);
        return makeSpan(builder);
    }

    public SpanBuilder withImage(Drawable drawable, int verticalAlignment) {
        ImageSpanBuilder builder = new ImageSpanBuilder(drawable, verticalAlignment);
        return makeSpan(builder);
    }

    public SpanBuilder withImage(Drawable drawable, String source, int verticalAlignment) {
        ImageSpanBuilder builder = new ImageSpanBuilder(drawable, source, verticalAlignment);
        return makeSpan(builder);
    }

    public SpanBuilder withImage(Context context, Uri uri) {
        ImageSpanBuilder builder = new ImageSpanBuilder(context, uri);
        return makeSpan(builder);
    }

    public SpanBuilder withImage(Context context, Uri uri, int verticalAlignment) {
        ImageSpanBuilder builder = new ImageSpanBuilder(context, uri, verticalAlignment);
        return makeSpan(builder);
    }

    public SpanBuilder withImage(Context context, int resourceId) {
        ImageSpanBuilder builder = new ImageSpanBuilder(context, resourceId);
        return makeSpan(builder);
    }

    public SpanBuilder withImage(Context context, int resourceId, int verticalAlignment) {
        ImageSpanBuilder builder = new ImageSpanBuilder(context, resourceId, verticalAlignment);
        return makeSpan(builder);
    }

    public SpanBuilder withXScale(float proportion) {
        ScaleXSpanBuilder builder = new ScaleXSpanBuilder(proportion);
        return makeSpan(builder);
    }

    public SpanBuilder withXScale(Parcel src) {
        ScaleXSpanBuilder builder = new ScaleXSpanBuilder(src);
        return makeSpan(builder);
    }

    public SpanBuilder withStyle(int style) {
        StyleSpanBuilder builder = new StyleSpanBuilder(style);
        return makeSpan(builder);
    }

    public SpanBuilder withStyle(Parcel src) {
        StyleSpanBuilder builder = new StyleSpanBuilder(src);
        return makeSpan(builder);
    }

    public SpanBuilder withTypeface(String family) {
        TypeFaceSpanBuilder builder = new TypeFaceSpanBuilder(family);
        return makeSpan(builder);
    }

    public SpanBuilder withTypeface(Parcel src) {
        TypeFaceSpanBuilder builder = new TypeFaceSpanBuilder(src);
        return makeSpan(builder);
    }


    /**
     * Subscript and superscript span skipped intentionally , test its behavior in future
     */


    public SpanBuilder withTextAppearance(Context context, int appearance) {
        TextAppearanceSpanBuilder builder = new TextAppearanceSpanBuilder(context, appearance);
        return makeSpan(builder);
    }

    public SpanBuilder withTextAppearance(Context context, int appearance, int colorList) {
        TextAppearanceSpanBuilder builder = new TextAppearanceSpanBuilder(context, appearance, colorList);
        return makeSpan(builder);
    }

    public SpanBuilder withTextAppearance(String family, int style, int size, ColorStateList color, ColorStateList linkColor) {
        TextAppearanceSpanBuilder builder = new TextAppearanceSpanBuilder(family, style, size, color, linkColor);
        return makeSpan(builder);
    }

    public SpanBuilder withTextAppearance(Parcel src) {
        TextAppearanceSpanBuilder builder = new TextAppearanceSpanBuilder(src);
        return makeSpan(builder);
    }

    /**
     * Rasterizer span skipped since its rasterizer is deprecated since api 21
     */

    public SpanBuilder withStrikethrough() {
        StrikeThroughSpanBuilder builder = new StrikeThroughSpanBuilder();
        return makeSpan(builder);
    }

    public SpanBuilder withStrikeThrough(Parcel src) {
        StrikeThroughSpanBuilder builder = new StrikeThroughSpanBuilder(src);
        return makeSpan(builder);
    }


    public SpanBuilder withSuggestion(Context context, String[] suggestions, int flags) {
        SuggestionSpanBuilder builder = new SuggestionSpanBuilder(context, suggestions, flags);
        return makeSpan(builder);
    }

    public SpanBuilder withSuggestion(Locale locale, String[] suggestions, int flags) {
        SuggestionSpanBuilder builder = new SuggestionSpanBuilder(locale, suggestions, flags);
        return makeSpan(builder);
    }

    public SpanBuilder withSuggestion(Context context, Locale locale, String[] suggestions,
                                      int flags, Class<?> notificationTargetClass) {
        SuggestionSpanBuilder builder = new SuggestionSpanBuilder(context, locale, suggestions,
                flags, notificationTargetClass);
        return makeSpan(builder);
    }

    public SpanBuilder withSuggestion(Parcel src) {
        SuggestionSpanBuilder builder = new SuggestionSpanBuilder(src);
        return makeSpan(builder);
    }

    public SpanBuilder withUnderline() {
        UnderlineSpanBuilder builder = new UnderlineSpanBuilder();
        return makeSpan(builder);
    }

    public SpanBuilder withUnderline(Parcel src) {
        UnderlineSpanBuilder builder = new UnderlineSpanBuilder(src);
        return makeSpan(builder);
    }

    public SpanBuilder withStandardAlignment(Layout.Alignment alignment) {
        AlignmentSpanStandardBuilder builder = new AlignmentSpanStandardBuilder(alignment);
        return makeSpan(builder);
    }

    public SpanBuilder withStandardAlignment(Parcel src) {
        AlignmentSpanStandardBuilder builder = new AlignmentSpanStandardBuilder(src);
        return makeSpan(builder);
    }

    public SpanBuilder withBullet() {
        BulletSpanBuilder builder = new BulletSpanBuilder();
        return makeSpan(builder);
    }

    public SpanBuilder withBullet(int gapWidth) {
        BulletSpanBuilder builder = new BulletSpanBuilder();
        return makeSpan(builder);
    }

    public SpanBuilder withBullet(int gapWidth, int color) {
        BulletSpanBuilder builder = new BulletSpanBuilder(gapWidth, color);
        return makeSpan(builder);
    }

    public SpanBuilder withBullet(Parcel src) {
        BulletSpanBuilder builder = new BulletSpanBuilder(src);
        return makeSpan(builder);
    }

    public SpanBuilder withDrawableMargin(Drawable d) {
        DrawableMarginSpanBuilder builder = new DrawableMarginSpanBuilder(d);
        return makeSpan(builder);
    }

    public SpanBuilder withDrawableMargin(Drawable d, int pad) {
        DrawableMarginSpanBuilder builder = new DrawableMarginSpanBuilder(d, pad);
        return makeSpan(builder);
    }

    public SpanBuilder withIcon(Bitmap b) {
        IconSpanBuilder builder = new IconSpanBuilder(b);
        return makeSpan(builder);
    }

    public SpanBuilder withIcon(Bitmap b, int pad) {
        IconSpanBuilder builder = new IconSpanBuilder(b, pad);
        return makeSpan(builder);
    }

    public SpanBuilder withLeadingStandard(int first, int rest) {
        LeadingMarginStandardSpanBuilder builder = new LeadingMarginStandardSpanBuilder(first, rest);
        return makeSpan(builder);
    }

    public SpanBuilder withLeadingStandard(int every) {
        LeadingMarginStandardSpanBuilder builder = new LeadingMarginStandardSpanBuilder(every);
        return makeSpan(builder);
    }

    public SpanBuilder withLeadingStandard(Parcel src) {
        LeadingMarginStandardSpanBuilder builder = new LeadingMarginStandardSpanBuilder(src);
        return makeSpan(builder);
    }

    public SpanBuilder withQuote() {
        QuoteSpanBuilder builder = new QuoteSpanBuilder();
        return makeSpan(builder);
    }

    public SpanBuilder withQuote(int color) {
        QuoteSpanBuilder builder = new QuoteSpanBuilder(color);
        return makeSpan(builder);
    }

    public SpanBuilder withQuote(Parcel src) {
        QuoteSpanBuilder builder = new QuoteSpanBuilder(src);
        return makeSpan(builder);
    }

    public SpanBuilder withIconMargin(Bitmap b) {
        IconMarginSpanBuilder builder = new IconMarginSpanBuilder(b);
        return makeSpan(builder);
    }

    public SpanBuilder withIconMargin(Bitmap b, int pad) {
        IconMarginSpanBuilder builder = new IconMarginSpanBuilder(b, pad);
        return makeSpan(builder);
    }

    public SpanBuilder withTabStopStandard(int where) {
        TabStopStandardSpanBuilder builder = new TabStopStandardSpanBuilder(where);
        return makeSpan(builder);
    }

    /**
     * Internal methods
     */

    public SpanBuilder withFlag(int flags) {
        this.flags = flags;
        return this;
    }


    public SpanBuilder resetFlags() {
        this.flags = Spanned.SPAN_INCLUSIVE_EXCLUSIVE;
        return this;
    }

    public SpanBuilder index(int startIndex, int endIndex) {
        if (startIndex > endIndex)
            throw new IllegalArgumentException("Start index cannot be higher than end index at: " +
                    "startIndex: " + startIndex + "endIndex: " + endIndex);
        this.startIndex = startIndex;
        this.endIndex = endIndex;
        return this;
    }

    public SpanBuilder resetIndex() {
        this.startIndex = 0;
        this.endIndex = text.length();
        return this;
    }

    public String getText() {
        return text;
    }

    public TextView getTextView() {
        return textView;
    }

    public int getStartIndex() {
        return startIndex;
    }

    public int getEndIndex() {
        return endIndex;
    }

    public int getFlags() {
        return flags;
    }

    public SpannableStringBuilder build() {
        return spannableBuilder;
    }

    /**
     * If want to use textView directly with view this method must be invoked at the beginning of a chain
     * in order to retain all functionality
     */
    public SpanBuilder withView(TextView textView) {
        this.textView = textView;
        return this;
    }

    public void buildWithTextView() {
        if (textView == null)
            throw new IllegalStateException("Cannot call buildWithTextView() without calling withView()");
        textView.setText(spannableBuilder);
    }


    private SpanProxy createProxy(int start, int end) {
        return new SpanProxy(this, text, start, end);
    }

    private SpanProxy getDefaultProxy() {
        return createProxy(startIndex, endIndex, flags);
    }

    private SpanProxy createProxy(int start, int end, int flags) {
        return new SpanProxy(this, text, start, end, flags);
    }

    private SpanBuilder makeSpan(SpanTypeBuilder builder) {
        return builder.make(getDefaultProxy());
    }

}

File: span-builder/src/main/java/com/maciega/bartosz/spanbuilder/spans/SpanBuilder.kt
package com.maciega.bartosz.spanbuilder.spans

import android.annotation.TargetApi
import android.content.Context
import android.content.res.ColorStateList
import android.graphics.Bitmap
import android.graphics.MaskFilter
import android.graphics.drawable.Drawable
import android.net.Uri
import android.os.Build
import android.support.annotation.ColorInt
import android.text.Layout
import android.text.SpannableStringBuilder
import android.text.Spanned
import android.text.method.LinkMovementMethod
import android.text.style.AbsoluteSizeSpan
import android.text.style.AlignmentSpan
import android.text.style.BackgroundColorSpan
import android.text.style.BulletSpan
import android.text.style.ClickableSpan
import android.text.style.DrawableMarginSpan
import android.text.style.ForegroundColorSpan
import android.text.style.IconMarginSpan
import android.text.style.ImageSpan
import android.text.style.LeadingMarginSpan
import android.text.style.LocaleSpan
import android.text.style.MaskFilterSpan
import android.text.style.QuoteSpan
import android.text.style.RelativeSizeSpan
import android.text.style.ScaleXSpan
import android.text.style.StrikethroughSpan
import android.text.style.StyleSpan
import android.text.style.SuggestionSpan
import android.text.style.TabStopSpan
import android.text.style.TextAppearanceSpan
import android.text.style.TypefaceSpan
import android.text.style.URLSpan
import android.text.style.UnderlineSpan
import android.view.View
import android.widget.TextView
import java.util.Locale

const val ZERO_START_INDEX = 0
const val DEFAULT_SPAN_FLAG = Spanned.SPAN_INCLUSIVE_EXCLUSIVE

inline fun TextView.withSpan(text: String, flags: Int = DEFAULT_SPAN_FLAG, buildFunc: SpanBuilder.(TextView) -> Unit) {
  val builder = SpanBuilder(text, flags)
  buildFunc.invoke(builder, this)
  this.text = builder.build()
}

class SpanBuilder constructor(text: String, private val initialFlags: Int) {

  constructor(text: String, flags: Int = DEFAULT_SPAN_FLAG, buildFunc: SpanBuilder.() -> Unit) : this(text, flags) {
    buildFunc.invoke(this)
  }

  private val innerSpanBuilder: SpannableStringBuilder = SpannableStringBuilder(text)
  private val textLength = text.length

  fun backgroundColor(@ColorInt color: Int, start: Int = ZERO_START_INDEX, end: Int = textLength, flags: Int = initialFlags) {
    innerSpanBuilder.setSpan(BackgroundColorSpan(color), start, end, flags)
  }

  fun foregroundColor(@ColorInt color: Int, start: Int = ZERO_START_INDEX, end: Int = textLength, flags: Int = initialFlags) {
    innerSpanBuilder.setSpan(ForegroundColorSpan(color), start, end, flags)
  }

  fun clickable(textView: TextView, start: Int = ZERO_START_INDEX, end: Int = textLength, flags: Int = initialFlags, listener: (View) -> Unit) {
    val clickableSpan = object : ClickableSpan() {
      override fun onClick(widget: View) {
        listener.invoke(widget)
      }
    }
    innerSpanBuilder.setSpan(clickableSpan, start, end, flags)
    textView.movementMethod = LinkMovementMethod.getInstance()
  }

  fun url(url: String, textView: TextView, start: Int = ZERO_START_INDEX, end: Int = textLength, flags: Int = initialFlags) {
    innerSpanBuilder.setSpan(URLSpan(url), start, end, flags)
    textView.movementMethod = LinkMovementMethod.getInstance()
  }

  fun mask(filter: MaskFilter, start: Int = ZERO_START_INDEX, end: Int = textLength, flags: Int = initialFlags) {
    innerSpanBuilder.setSpan(MaskFilterSpan(filter), start, end, flags)
  }

  fun absoluteSize(size: Int, dip: Boolean = false, start: Int = ZERO_START_INDEX, end: Int = textLength, flags: Int = initialFlags) {
    innerSpanBuilder.setSpan(AbsoluteSizeSpan(size, dip), start, end, flags)
  }

  @TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR1)
  fun locale(locale: Locale, start: Int = ZERO_START_INDEX, end: Int = textLength, flags: Int = initialFlags) {
    innerSpanBuilder.setSpan(LocaleSpan(locale), start, end, flags)
  }

  fun relative(proportion: Float, start: Int = ZERO_START_INDEX, end: Int = textLength, flags: Int = initialFlags) {
    innerSpanBuilder.setSpan(RelativeSizeSpan(proportion), start, end, flags)
  }

  fun image(context: Context, bitmap: Bitmap, verticalAlignment: Int? = null, start: Int = ZERO_START_INDEX, end: Int = textLength, flags: Int = initialFlags) {
    val imageSpan = if (verticalAlignment != null) ImageSpan(context, bitmap, verticalAlignment) else ImageSpan(context, bitmap)
    innerSpanBuilder.setSpan(imageSpan, start, end, flags)
  }

  fun image(context: Context, uri: Uri, verticalAlignment: Int? = null, start: Int = ZERO_START_INDEX, end: Int = textLength, flags: Int = initialFlags) {
    val imageSpan = if (verticalAlignment != null) ImageSpan(context, uri, verticalAlignment) else ImageSpan(context, uri)
    innerSpanBuilder.setSpan(imageSpan, start, end, flags)
  }

  fun image(context: Context, resourceId: Int, verticalAlignment: Int? = null, start: Int = ZERO_START_INDEX, end: Int = textLength, flags: Int = initialFlags) {
    val imageSpan = if (verticalAlignment != null) ImageSpan(context, resourceId, verticalAlignment) else ImageSpan(context, resourceId)
    innerSpanBuilder.setSpan(imageSpan, start, end, flags)
  }

  fun image(drawable: Drawable, source: String? = null, verticalAlignment: Int? = null, start: Int = ZERO_START_INDEX, end: Int = textLength, flags: Int = initialFlags) {
    val imageSpan = when (source) {
      null -> if (verticalAlignment == null) ImageSpan(drawable) else ImageSpan(drawable, verticalAlignment)
      else -> if (verticalAlignment == null) ImageSpan(drawable, source) else ImageSpan(drawable, source, verticalAlignment)
    }
    innerSpanBuilder.setSpan(imageSpan, start, end, flags)
  }

  fun scaleX(proportion: Float, start: Int = ZERO_START_INDEX, end: Int = textLength, flags: Int = initialFlags) {
    innerSpanBuilder.setSpan(ScaleXSpan(proportion), start, end, flags)
  }

  fun style(style: Int, start: Int = ZERO_START_INDEX, end: Int = textLength, flags: Int = initialFlags) {
    innerSpanBuilder.setSpan(StyleSpan(style), start, end, flags)
  }

  fun typeface(family: String, start: Int = ZERO_START_INDEX, end: Int = textLength, flags: Int = initialFlags) {
    innerSpanBuilder.setSpan(TypefaceSpan(family), start, end, flags)
  }

  fun textAppearance(context: Context, appearance: Int, colorList: Int = -1, start: Int = ZERO_START_INDEX, end: Int = textLength, flags: Int = initialFlags) {
    innerSpanBuilder.setSpan(TextAppearanceSpan(context, appearance, colorList), start, end, flags)
  }

  fun textAppearance(
    family: String,
    style: Int,
    size: Int,
    color: ColorStateList,
    linkColor: ColorStateList,
    start: Int = ZERO_START_INDEX,
    end: Int = textLength,
    flags: Int = initialFlags
  ) {
    innerSpanBuilder.setSpan(TextAppearanceSpan(family, style, size, color, linkColor), start, end, flags)
  }

  fun strikeThrough(start: Int = ZERO_START_INDEX, end: Int = textLength, flags: Int = initialFlags) {
    innerSpanBuilder.setSpan(StrikethroughSpan(), start, end, flags)
  }

  fun suggestion(context: Context, suggestions: Array<String>, spanFlags: Int, start: Int = ZERO_START_INDEX, end: Int = textLength, flags: Int = initialFlags) {
    innerSpanBuilder.setSpan(SuggestionSpan(context, suggestions, spanFlags), start, end, flags)
  }

  fun suggestion(locale: Locale, suggestions: Array<String>, spanFlags: Int, start: Int = ZERO_START_INDEX, end: Int = textLength, flags: Int = initialFlags) {
    innerSpanBuilder.setSpan(SuggestionSpan(locale, suggestions, spanFlags), start, end, flags)
  }

  fun suggestion(
    context: Context? = null,
    locale: Locale? = null,
    suggestions: Array<String>,
    spanFlags: Int,
    notificationTargetClass: Class<Any>? = null,
    start: Int = ZERO_START_INDEX,
    end: Int = textLength,
    flags: Int = initialFlags
  ) {
    innerSpanBuilder.setSpan(SuggestionSpan(context, locale, suggestions, spanFlags, notificationTargetClass), start, end, flags)
  }

  fun underline(start: Int = ZERO_START_INDEX, end: Int = textLength, flags: Int = initialFlags) {
    innerSpanBuilder.setSpan(UnderlineSpan(), start, end, flags)
  }

  fun standardAlignment(alignment: Layout.Alignment, start: Int = ZERO_START_INDEX, end: Int = textLength, flags: Int = initialFlags) {
    innerSpanBuilder.setSpan(AlignmentSpan.Standard(alignment), start, end, flags)
  }

  fun bullet(gapWidth: Int? = null, @ColorInt color: Int? = null, start: Int = ZERO_START_INDEX, end: Int = textLength, flags: Int = initialFlags) {
    val bulletSpan = when {
      gapWidth != null -> if (color != null) BulletSpan(gapWidth, color) else BulletSpan(gapWidth)
      color == null -> BulletSpan()
      else -> throw InvalidSpanArgumentsException("Invalid arguments, gapWidth cannot be null when color is not null")
    }
    innerSpanBuilder.setSpan(bulletSpan, start, end, flags)
  }

  fun drawableMargin(drawable: Drawable, padding: Int? = null, start: Int = ZERO_START_INDEX, end: Int = textLength, flags: Int = initialFlags) {
    val drawableMarginSpan = if (padding == null) {
      DrawableMarginSpan(drawable)
    } else {
      DrawableMarginSpan(drawable, padding)
    }
    innerSpanBuilder.setSpan(drawableMarginSpan, start, end, flags)
  }

  fun iconMargin(bitmap: Bitmap, padding: Int? = null, start: Int = ZERO_START_INDEX, end: Int = textLength, flags: Int = initialFlags) {
    val iconMarginSpan = if (padding == null) {
      IconMarginSpan(bitmap)
    } else {
      IconMarginSpan(bitmap, padding)
    }
    innerSpanBuilder.setSpan(iconMarginSpan, start, end, flags)
  }

  fun leadingMarginStandard(first: Int? = null, rest: Int? = null, every: Int? = null, start: Int = ZERO_START_INDEX, end: Int = textLength, flags: Int = initialFlags) {
    val leadingMarginSpan = when {
      every != null -> LeadingMarginSpan.Standard(every)
      rest != null && first != null -> LeadingMarginSpan.Standard(first, rest)
      else -> throw InvalidSpanArgumentsException("Invalid arguments, must to declare every, or both first and rest")
    }
    innerSpanBuilder.setSpan(leadingMarginSpan, start, end, flags)
  }

  fun quoteSpan(@ColorInt color: Int? = null, start: Int = ZERO_START_INDEX, end: Int = textLength, flags: Int = initialFlags) {
    val quoteSpan = when (color) {
      null -> QuoteSpan()
      else -> QuoteSpan(color)
    }
    innerSpanBuilder.setSpan(quoteSpan, start, end, flags)
  }

  fun tabStopStandard(offset: Int, start: Int = ZERO_START_INDEX, end: Int = textLength, flags: Int = initialFlags) {
    innerSpanBuilder.setSpan(TabStopSpan.Standard(offset), start, end, flags)
  }

  fun build() = innerSpanBuilder
}